src/Type/Constant/ConstantArrayType.php

```diff
@@ -1012,6 +1006,120 @@ public function sliceArray(Type $offsetType, Type $lengthType, TrinaryLogic $pre
 		return $builder->getArray();
 	}
 
+	public function spliceArray(Type $offsetType, Type $lengthType, Type $replacementType): Type
+	{
+		$keyTypesCount = count($this->keyTypes);
+
+		$offset = $offsetType instanceof ConstantIntegerType ? $offsetType->getValue() : null;
+
+		if ($lengthType instanceof ConstantIntegerType) {
+			$length = $lengthType->getValue();
+		} elseif ($lengthType->isNull()->yes()) {
+			$length = $keyTypesCount;
+		} else {
+			$length = null;
+		}
+
+		if ($offset === null || $length === null) {
+			return $this->degradeToGeneralArray()
+				->spliceArray($offsetType, $lengthType, $replacementType);
+		}
+
+		if ($keyTypesCount + $offset <= 0) {
+			// A negative offset cannot reach left outside the array twice
+			$offset = 0;
+		}
+
+		if ($offset < 0) {
+			/*
+			 * Transforms the problem with the negative offset in one with a positive offset using array reversion.
+			 * The reason is belows handling of optional keys which works only from left to right.
+			 *
+			 * e.g.
+			 * array{a: 0, b: 1, c: 2, d: 3, e: 4}
+			 * with offset -4, length 2 and replacement array{17, 19} (which would be spliced to array{a: 0, 0: 17, 1: 19, d: 3, e: 4})
+			 *
+			 * is transformed via reversion to
+			 *
+			 * array{e: 4, d: 3, c: 2, b: 1, a: 0}
+			 * with offset 2 and length 2 (which will be spliced to array{e: 4, d: 3, 1: 19, 0: 17, a: 0} and then reversed again)
+			 */
+			$offset *= -1;
+			$reversedLength = min($length, $offset);
+			$reversedOffset = $offset - $reversedLength;
```

Comment:
these operations look like they could be expressed on the type-level directly.

but as you said, maybe this would be too much and we leave it less abstract for now and see whether people report new issues

Reply:
Can we do such calculations too, can you show me an example maybe? I'm not too confident that I can make it work, but might be also a good follow-up.

Reply:
maybe https://github.com/phpstan/phpstan-src/blob/e6d9f0c3ec251ea6fad604d3f97e02bedb4d7f42/src/Type/Php/ArraySumFunctionDynamicReturnTypeExtension.php#L36-L56
or `InitializerExprTypeResolver::get(Minus|Plus|...)Type`

Reply:
can't figure out how to do this now unfortunately, there's also no access to scope in the type. but it's way simpler because non-optional keys don't have to be considered and there's a chance it might work somehow I suppose. maybe still a good follow-up :)